lightning/src/ln/channel.rs

```diff
@@ -1598,6 +1617,84 @@ impl<Signer: ChannelSigner> ChannelContext<Signer> {
 		stats
 	}
 
+
+	/// Returns information on all pending inbound HTLCs.
+	pub fn get_pending_inbound_htlc_details(&self) -> Vec<HTLCDetails> {
+		let mut inbound_details = Vec::new();
+		let htlc_success_dust_limit = if self.get_channel_type().supports_anchors_zero_fee_htlc_tx() {
+			0
+		} else {
+			let dust_buffer_feerate = self.get_dust_buffer_feerate(None) as u64;
+			dust_buffer_feerate * htlc_success_tx_weight(self.get_channel_type()) / 1000
+		};
+		let holder_dust_limit_success_sat = htlc_success_dust_limit + self.holder_dust_limit_satoshis;
+		for ref htlc in self.pending_inbound_htlcs.iter() {
```

Comment:
I'm not sure about the rationale for not including claimed/failed holding cell HTLCs. It seems like we should include them (and indicate whether they're successful or failed), and maybe add a note in `ChannelDetails::balance_msat` about the considerations from the PR description, if necessary?

Reply:
Thinking about it more, I think there's two approaches we can take to pending HTLCs and balances if we want a current view independent from commitment transactions:
- Being as conservative as possible and considering as many pending HTLCs as encumbering the balance as possible. The pending HTLCs will therefore include both added holding cell HTLCs on one end and HTLCs being partially resolved on the other end. These pending HTLCs will subtract from the balance. Resolutions (claimed/failed) in the holding cell won't be considered.
- Being as aggressive as possible and considering as many HTLCs as resolved as possible and removing them from the pending HTLCs and adjusting the balance appropriately. The pending HTLCs will therefore not include HTLCs that are in a partially resolved state or for which the resolution is in the holding cell. One question here is around partially failed HTLCs where one party can still broadcast a previous state and fulfill the HTLC.

Currently, we are very close to being fully consistent with the former, conservative option, minus the inclusion of inbound-claimed HTLCs in `AvailableBalances::balance_msat` due to https://github.com/lightningdevkit/rust-lightning/pull/1268/files. 

However, the root cause of the underflow that was the motivation of that PR does not exist anymore. At the time of the PR, the amount in `send_htlc` was [verified against the CommitmentStats](https://github.com/lightningdevkit/rust-lightning/blob/c931380cbc76bfa0bab6b039f0de15474faddef9/lightning/src/ln/channel.rs#L5010-L5013) (and this commitment transaction resolves the pending HTLC in the `InboundHTLCState::LocalRemoved(InboundHTLCRemovalReason::Fulfill(_))` state). Currently, we're [verifying](https://github.com/lightningdevkit/rust-lightning/blob/4938be6ad715d3474c6bd13c914fd05490e6a649/lightning/src/ln/channel.rs#L5166-L5169) against `AvailableBalances::next_outbound_htlc_limit_msat` which [does not resolve such pending HTLCs](https://github.com/lightningdevkit/rust-lightning/blob/4938be6ad715d3474c6bd13c914fd05490e6a649/lightning/src/ln/channel.rs#L1622-L1739). I copied and reran the test in https://github.com/lightningdevkit/rust-lightning/pull/1268 from the main branch and it is not possible to send an HTLC back anymore and cause an underflow (sending the return payment fails with `Cannot send more than our next-HTLC maximum`). This is likely why the test has been removed.

Therefore, I want to propose the conservative approach and remove the inclusion of inbound-claimed HTLCs in `AvailableBalances::balance_msat`. This will also cause the sum of the reserves, inbound capacity, outbound capacity and pending HTLCs to be equal to the channel size when the reserves are met. What do you think @valentinewallace @TheBlueMatt?